ulib/experimental/Steel.Channel.Simplex.fst

```diff
@@ -273,8 +273,9 @@ let gather_r (#p:sprot) (r:ref chan_val) (v:chan_val)
     (fun _ -> pts_to r full_perm v `star` in_state_slprop p v)
   = let v' = witness_h_exists () in
     Steel.Utils.higher_ref_pts_to_injective_eq r v _;
-    H.gather r;
-    rewrite_context()
+    H.gather #_ #half #half #v #v r;
+    change_slprop (pts_to r (sum_perm half half) v) (pts_to r full_perm v) (fun _ -> ());
+    change_slprop (in_state_slprop p v') (in_state_slprop p v) (fun _ -> ())
```

Comment:
Any ideas why these changes? Are the constraints different and we regressed a bit in solving them?

Reply:
This rewrite_context has always been a bit finicky, since it relies on SMT solving to prove the slprop-equivalence, it needs the shape of both slprops to match exactly.
Previously, we were "lucky" in a few cases: the inferred postcondition of the computation was matching the annotation, and no rewriting was performed (when discharging separation logic constraints, we first attempt reflexivity before jumping into the core of canon which performs AC-rewritings).

Now that we are always adding frames, even though they are often resolved to emp, shapes rarely match, which mean we have to fall back on piecewise rewritings more often